File: app/src/test/java/com/matchthree/game/engine/ScorerTest.kt
package com.matchthree.game.engine

import com.matchthree.game.model.Position
import org.junit.Assert.assertEquals
import org.junit.Test

class ScorerTest {

    @Test
    fun `base rate is 10 points per unique gem at depth 1`() {
        assertEquals(30, Scorer.roundScore(setOf(pos(0, 0), pos(0, 1), pos(0, 2)), 1))
        assertEquals(50, Scorer.roundScore(setOf(pos(1, 0), pos(1, 1), pos(1, 2), pos(2, 2), pos(3, 2)), 1))
    }

    @Test
    fun `cascade multiplier scales linearly and uncapped by depth`() {
        val cells = setOf(pos(0, 0), pos(0, 1), pos(0, 2))
        assertEquals(30, Scorer.roundScore(cells, 1))
        assertEquals(60, Scorer.roundScore(cells, 2))
        assertEquals(90, Scorer.roundScore(cells, 3))
        assertEquals(300, Scorer.roundScore(cells, 10))
    }

    @Test
    fun `overlapping clear regions are counted once (unique-cell scoring)`() {
        // A row clear and a column clear sharing the crossing cell (2,2):
        // 5 unique cells, not 6.
        val rowClear = listOf(pos(2, 0), pos(2, 1), pos(2, 2))
        val columnClear = listOf(pos(0, 2), pos(1, 2), pos(2, 2))
        val combined = rowClear + columnClear
        assertEquals(5, combined.toSet().size)
        assertEquals(50, Scorer.roundScore(combined, 1))
    }

    @Test
    fun `duplicates passed to a single round are deduped`() {
        assertEquals(20, Scorer.roundScore(listOf(pos(0, 0), pos(0, 1), pos(0, 0)), 1))
    }

    @Test
    fun `total score sums every round`() {
        assertEquals(0, Scorer.totalScore(emptyList()))
        assertEquals(30, Scorer.totalScore(listOf(30)))
        assertEquals(30 + 60 + 90, Scorer.totalScore(listOf(30, 60, 90)))
    }

    @Test(expected = IllegalArgumentException::class)
    fun `depth below 1 is rejected`() {
        Scorer.roundScore(emptySet(), 0)
    }

    private fun pos(row: Int, col: Int) = Position(row, col)
}

File: Tests/ScorerTest.cs
using MatchThree.Engine.Model;
using MatchThree.Engine.Rules;
using NUnit.Framework;

namespace MatchThree.Engine.Tests;

public class ScorerTest
{
    [Test]
    public void BaseRateIs10PointsPerUniqueGemAtDepth1()
    {
        Assert.That(Scorer.RoundScore(new[] { Pos(0, 0), Pos(0, 1), Pos(0, 2) }, 1), Is.EqualTo(30));
        Assert.That(Scorer.RoundScore(
            new[] { Pos(1, 0), Pos(1, 1), Pos(1, 2), Pos(2, 2), Pos(3, 2) }, 1), Is.EqualTo(50));
    }

    [Test]
    public void CascadeMultiplierScalesLinearlyAndUncappedByDepth()
    {
        var cells = new[] { Pos(0, 0), Pos(0, 1), Pos(0, 2) };
        Assert.That(Scorer.RoundScore(cells, 1), Is.EqualTo(30));
        Assert.That(Scorer.RoundScore(cells, 2), Is.EqualTo(60));
        Assert.That(Scorer.RoundScore(cells, 3), Is.EqualTo(90));
        Assert.That(Scorer.RoundScore(cells, 10), Is.EqualTo(300));
    }

    [Test]
    public void OverlappingClearRegionsAreCountedOnce()
    {
        var rowClear = new[] { Pos(2, 0), Pos(2, 1), Pos(2, 2) };
        var columnClear = new[] { Pos(0, 2), Pos(1, 2), Pos(2, 2) };
        var combined = rowClear.Concat(columnClear).ToList();
        Assert.That(combined.Distinct().ToList(), Has.Count.EqualTo(5));
        Assert.That(Scorer.RoundScore(combined, 1), Is.EqualTo(50));
    }

    [Test]
    public void DuplicatesPassedToASingleRoundAreDeduped()
    {
        Assert.That(Scorer.RoundScore(new[] { Pos(0, 0), Pos(0, 1), Pos(0, 0) }, 1), Is.EqualTo(20));
    }

    [Test]
    public void TotalScoreSumsEveryRound()
    {
        Assert.That(Scorer.TotalScore(Array.Empty<int>()), Is.EqualTo(0));
        Assert.That(Scorer.TotalScore(new[] { 30 }), Is.EqualTo(30));
        Assert.That(Scorer.TotalScore(new[] { 30, 60, 90 }), Is.EqualTo(30 + 60 + 90));
    }

    [Test]
    public void DepthBelow1IsRejected()
    {
        Assert.Throws<ArgumentException>(() => Scorer.RoundScore(Array.Empty<Position>(), 0));
    }

    private static Position Pos(int row, int col) => new(row, col);
}